sram.c

```diff
@@ -0,0 +1,96 @@
+#include "sram.h"
+#include "spi.h"
+#include "nuts_bolts.h"
+
+/* Least and most significant bytes of x, of type uint16_t */
+#define LSB(x) (x & 0x00FF)
+#define MSB(x) ((x & 0xFF00) >> 8)
+
+#define MODE_MASK 0xC0
+#define MODE_IDX  6U
+
+/* SPI insructions */
+enum instruction_e {
+  READ = 0x3,
+  WRITE = 0x2,
+  EDIO = 0x3B,
+  EQIO = 0x38,
+  RSTIO = 0xFF,
+  RDMR = 0x5,
+  WRMR = 0x1
+};
+
+void sram_init()
+{
+  /* Configure SCS pin as output */
+  SCS_SRAM_DDR |= (1 << SCS_SRAM_DDR_PIN);
+
+  /* Set SCS pin to high. The CS for this chip is active low */
+  SCS_SRAM_PORT &= ~(1 << SCS_SRAM_PIN);
+
+  /* SCK resting state is 0. Clock data on rising edge */
+  spi_set_mode(0, 0);
+
+  sram_set_mode(BYTE_MODE);
+}
+
+uint8_t sram_read_mode()
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {RDMR, 0xFF};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+
+  return (data_in[1] & MODE_MASK) >> MODE_IDX;
+}
+
+void sram_set_mode(enum sram_mode_e mode)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[2] = {WRMR, mode << MODE_IDX};
+  uint8_t data_in[2];
+
+  bit_false(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+  spi_transact_array(data_out, data_in, 2);
+  bit_true(SCS_SRAM_PORT, 1 << SCS_SRAM_PIN);
+}
+
+uint8_t sram_read_byte(uint16_t addr)
+{
+  spi_set_mode(0, 0);
+
+  uint8_t data_out[4] = {READ,
+                         MSB(addr),
+                         LSB(addr),
+                         0xFF};
```

Comment:
I was thinking of a function that parameterizes these two values and returns something, and then `read` and `write` are each 1-liners that call the helper:

```c
uint8_t sram_helper(uint16_t addr, uint8_t val, uint8_t op) {
    /* Performs the operation (READ or WRITE) on the given address with the
       given value, and returns the result. */
    spi_set_mode(0, 0);
    uint8_t data_out[4] = {op, MSB(addr), LSB(addr), val};
    ...
    return data_in[3];
}

uint8_t sram_read_byte(uint16_t addr) {
    return sram_helper(addr, 0xFF, READ);
}

void sram_write_byte(uint16_t addr, uint8_t val) {
    /* Ignore the return value. */
    sram_helper(addr, val, WRITE);
}
```

I don't know if this is actually an improvement, though.